R/util.R

```diff
@@ -0,0 +1,78 @@
+#' Determine the status of parallel processing. If cores are requested,
+#' set up the cores. If cores are registered, keep using those.
+#' @param parallel logical or integer: whether to use pre-existing parallel
+#'   workers, or how many to create and use. Default `TRUE`.
+#' @param silent logical: give warnings or updates in the computation?
+#'   Default `FALSE`.
+#' @keywords internal
+#' @return list (ncores, cluster) ncores is the number of cores and
+#'   cluster object is a parallel::makeCluster(parallel) object or FALSE
+#'   if there is no cluster.
+setupParallel <- function(parallel, silent = FALSE) {
```

Comment:
Is it possible to put these functions in separate util_....R files? Or add the `printPretty` later here?
It seems to me that `setupParallel` and `closeCoresOnExit` could already be merged into main, but `printPretty` still needs some work or there's still some possibility-thinking around it.

Reply:
It is an internal function that just formats some printing, why does it matter if it is merged in?
We will have too many util function files in the end, so I started gathering the smaller ones in util.R.

Reply:
Printing is tricky. We should probably have a discussion about it (channel in inferno?). There are different kinds of messages, different functions have different needs, and we must also consider R's general message policies. For example:
- `metadatatemplate()` emits many messages that can be quite crucial to the user, and it's only supposed to be used once at the beginning. Maybe the output should go both to `stdout` but also to some file (optional?), so the user can review it if they close the R session?
- Many functions that can use parallel processing may output information about the number of cores used etc. Maybe these should be printed by default, unless the user wraps them in `suppressMessages()`? That's how some other packages handle them. At the same time they can just be distracting to users who just want to do calculations. Ah, difficult decisions... See further comment in an issue that I'm going to open soon.
- `learn()` need a very different handling. It's essential to output to different log files for each core, for debugging – it'll be useful when users start finding issues, we can ask them to share their core-log files. It also makes sense to keep the user updated on estimated remaining time and so on. An option that should be added is to save the main output also to a log-file, maybe one with the same name as the output directory (it can't go in the output directory, though, because that's created after the messaging starts...). I'll try to add this extra log to it.

Reply:
Yes to all of the above, but I'm not trying to solve all of these things at once, I'm just making the printing a little nicer for my small internal functions.